File: _old/backend/billing/schemas.py
"""Pydantic models for the account, billing, and dashboard endpoints."""

from __future__ import annotations

from typing import Any, Literal, Optional

from pydantic import BaseModel, EmailStr, Field


# ---------------------------------------------------------------------------
# Auth
# ---------------------------------------------------------------------------

class SignupRequest(BaseModel):
    email: EmailStr
    name: str = Field(default="", max_length=120)
    # Length beats character-class rules; 8 is the floor.
    password: str = Field(..., min_length=8, max_length=200)
    # Optional and never validated here: a mistyped code costs the signup
    # nothing, it just goes unattributed.
    referral_code: str = Field(default="", max_length=32)


class ReferralInviteRequest(BaseModel):
    email: EmailStr


class LoginRequest(BaseModel):
    email: EmailStr
    password: str = Field(..., min_length=1, max_length=200)


class FirebaseAuthRequest(BaseModel):
    """
    The one-time exchange: a Firebase ID token for a session cookie.

    ``max_length`` is generous because an ID token carrying several provider
    identities and custom claims runs long, but it is bounded — an unbounded
    string here is a free way to make the server do RSA work on megabytes.
    """

    id_token: str = Field(..., min_length=1, max_length=8192)
    # Only read when this exchange ends up *creating* an account; a returning
    # user cannot be referred again. Unvalidated on purpose: a mistyped code
    # costs the sign-in nothing, it just goes unattributed.
    referral_code: str = Field(default="", max_length=32)


class CustomerResponse(BaseModel):
    id: str
    email: str
    name: str
    created_at: str


# ---------------------------------------------------------------------------
# Billing
# ---------------------------------------------------------------------------

class PlanResponse(BaseModel):
    id: str
    name: str
    price_cents: int
    price_display: str
    currency: str
    interval: str
    daily_credits: int
    tagline: str


class PricingResponse(BaseModel):
    currency: str
    plans: list[PlanResponse]
    yearly_savings_cents: int
    yearly_savings_display: str
    yearly_discount_percent: int
    yearly_effective_monthly_display: str
    trial_days: int


class SubscriptionResponse(BaseModel):
    plan_id: Optional[str] = None
    plan_name: Optional[str] = None
    status: str
    is_entitled: bool
    current_period_end: Optional[str] = None
    daily_credits: Optional[int] = None
    can_start_trial: bool = False


class CheckoutRequest(BaseModel):
    plan_id: Literal["monthly", "yearly"]


class CheckoutResponse(BaseModel):
    checkout_url: str
    reference: str
    provider: str
    requires_manual_confirmation: bool
    message: str


class InvoiceResponse(BaseModel):
    id: str
    plan_id: str
    amount_cents: int
    currency: str
    status: str
    issued_at: str
    paid_at: Optional[str] = None


# ---------------------------------------------------------------------------
# Dashboard — API keys
# ---------------------------------------------------------------------------

class CreateKeyRequest(BaseModel):
    label: str = Field(default="", max_length=60)


class CreatedKeyResponse(BaseModel):
    api_key: str
    key_id: str
    key_prefix: str
    message: str = "Copy this key now — it is not stored and cannot be shown again."


class DashboardKey(BaseModel):
    id: str
    key_prefix: str
    label: str
    created_at: str
    is_active: int


class DashboardResponse(BaseModel):
    customer: CustomerResponse
    subscription: SubscriptionResponse
    keys: list[DashboardKey]
    usage: dict[str, Any]

File: node-backend/src/billing/schemas.ts
/**
 * Zod schemas for the account, billing, and dashboard endpoints.
 *
 * Port of `backend/billing/schemas.py`.
 *
 * Pydantic did two jobs in the Python build: it validated incoming request
 * bodies, and it declared the *response* shape so FastAPI could both serialise
 * and document it. Zod covers the first job directly. For the second, the
 * response models become TypeScript interfaces — the compiler checks the shape
 * at build time, which is where the cost of getting it wrong belongs, and
 * nothing has to run at request time to reshape an object that is already
 * correct.
 *
 * `EmailStr` becomes `z.string().email()`, which applies the same basic
 * syntactic check. Pydantic's `email-validator` additionally normalises and can
 * check deliverability; neither was relied on here, and the real address policy
 * lives in `identity.ts` where it always did.
 */

import { z } from "zod";

// ---------------------------------------------------------------------------
// Auth
// ---------------------------------------------------------------------------

export const signupSchema = z.object({
  email: z.string().email(),
  name: z.string().max(120).default(""),
  // Length beats character-class rules; 8 is the floor.
  password: z.string().min(8).max(200),
  // Optional and never validated here: a mistyped code costs the signup
  // nothing, it just goes unattributed.
  referral_code: z.string().max(32).default(""),
});
export type SignupRequest = z.infer<typeof signupSchema>;

export const referralInviteSchema = z.object({
  email: z.string().email(),
});
export type ReferralInviteRequest = z.infer<typeof referralInviteSchema>;

export const loginSchema = z.object({
  email: z.string().email(),
  password: z.string().min(1).max(200),
});
export type LoginRequest = z.infer<typeof loginSchema>;

/**
 * The one-time exchange: a Firebase ID token for a session cookie.
 *
 * `max` is generous because an ID token carrying several provider identities
 * and custom claims runs long, but it is bounded — an unbounded string here is
 * a free way to make the server do RSA work on megabytes.
 */
export const firebaseAuthSchema = z.object({
  id_token: z.string().min(1).max(8192),
  // Only read when this exchange ends up *creating* an account; a returning
  // user cannot be referred again. Unvalidated on purpose: a mistyped code
  // costs the sign-in nothing, it just goes unattributed.
  referral_code: z.string().max(32).default(""),
});
export type FirebaseAuthRequest = z.infer<typeof firebaseAuthSchema>;

export interface CustomerResponse {
  id: string;
  email: string;
  name: string;
  created_at: string;
}

// ---------------------------------------------------------------------------
// Billing
// ---------------------------------------------------------------------------

export interface PlanResponse {
  id: string;
  name: string;
  price_cents: number;
  price_display: string;
  currency: string;
  interval: string;
  daily_credits: number;
  tagline: string;
}

export interface PricingResponse {
  currency: string;
  plans: PlanResponse[];
  yearly_savings_cents: number;
  yearly_savings_display: string;
  yearly_discount_percent: number;
  yearly_effective_monthly_display: string;
  trial_days: number;
}

export interface SubscriptionResponse {
  plan_id?: string | null;
  plan_name?: string | null;
  status: string;
  is_entitled: boolean;
  current_period_end?: string | null;
  daily_credits?: number | null;
  can_start_trial: boolean;
}

export const checkoutSchema = z.object({
  plan_id: z.enum(["monthly", "yearly"]),
});
export type CheckoutRequest = z.infer<typeof checkoutSchema>;

export interface CheckoutResponse {
  checkout_url: string;
  reference: string;
  provider: string;
  requires_manual_confirmation: boolean;
  message: string;
}

export interface InvoiceResponse {
  id: string;
  plan_id: string;
  amount_cents: number;
  currency: string;
  status: string;
  issued_at: string;
  paid_at?: string | null;
}

// ---------------------------------------------------------------------------
// Dashboard — API keys
// ---------------------------------------------------------------------------

export const createKeySchema = z.object({
  label: z.string().max(60).default(""),
});
export type CreateKeyRequest = z.infer<typeof createKeySchema>;

export interface CreatedKeyResponse {
  api_key: string;
  key_id: string;
  key_prefix: string;
  message: string;
}

/** The default `message` Pydantic supplied for {@link CreatedKeyResponse}. */
export const CREATED_KEY_MESSAGE =
  "Copy this key now — it is not stored and cannot be shown again.";

export interface DashboardKey {
  id: string;
  key_prefix: string;
  label: string;
  created_at: string;
  is_active: number;
}

export interface DashboardResponse {
  customer: CustomerResponse;
  subscription: SubscriptionResponse;
  keys: DashboardKey[];
  usage: Record<string, unknown>;
}
